[PyTorch] Redirect c10::optional to std::optional

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101995

We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @ezyang @gchanan @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @EikanWang @kiukchung @d4l3k @lucasllc

swolchok committed:
[PyTorch] Redirect c10::optional to std::optional

We have C++17 now!

I am intentionally dropping the c10::optional<c10::ArrayRef> size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use optional<ArrayRef> in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the c10::optional<c10::ArrayRef> size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use optional<ArrayRef> in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

[ghstack-poisoned]

albanD commented on c10/test/util/optional_test.cpp:
```diff
@@ -56,13 +58,6 @@ using OptionalTypes = ::testing::Types<
     // Non-trivial destructor.
     std::string>;
 
-// This assert is also in Optional.cpp; including here too to make it
```
cc @ezyang if you have more context about this

ezyang replied:
This is @swolchok's optimization, he knows ;)

albanD reviewed (commented):
Code change sounds fine, will let @malfet take the final look!

malfet commented on aten/src/ATen/core/ATen_fwd.h:
```diff
@@ -4,8 +4,6 @@
 // Forward declarations of core ATen types used in dispatch functions
 namespace c10 {
 
-template<typename T>
-class optional;
```
Should we just say something like `using optional = std::optional;` and kill `c10/util/Optional` header?

swolchok replied:
that'd be fine with me, but I've had to leave in `#include <c10/macros/Macros.h>` in `c10/util/Optional.h` for now because lots of files seem to depend on pulling in the `Macros.h` namespace shenanigans through `Optional.h` :(

Skylion007 replied:
Agre with @malfet . This is how llvm handled their std::optional conversion.

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the c10::optional<c10::ArrayRef> size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use optional<ArrayRef> in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the c10::optional<c10::ArrayRef> size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use optional<ArrayRef> in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

swolchok committed:
drop static_asserts in Optional.cpp because some random platform's std::optional doesn't pass them on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the c10::optional<c10::ArrayRef> size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use optional<ArrayRef> in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

swolchok commented:
been pulling my hair out over why `optTypeMetaToScalarType` was getting found before and now it requires `at::` or `c10::` in front. The answer is [ADL](https://en.cppreference.com/w/cpp/language/adl)! We will just have to fix all these oversights manually.

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the c10::optional<c10::ArrayRef> size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use optional<ArrayRef> in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the c10::optional<c10::ArrayRef> size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use optional<ArrayRef> in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the c10::optional<c10::ArrayRef> size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use optional<ArrayRef> in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the c10::optional<c10::ArrayRef> size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use optional<ArrayRef> in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the c10::optional<c10::ArrayRef> size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use optional<ArrayRef> in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the c10::optional<c10::ArrayRef> size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use optional<ArrayRef> in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

swolchok commented:
A bunch of functorch tests and some inductor tests are failing on this PR. AFAICT these particular tests don't run in fbcode. Any way I can figure out what's going on? I do have a stack trace blaming to a TORCH_INTERNAL_ASSERT that really shouldn't be firing; attempting a blind fix now but if that fails then I'll probably need some help.

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the c10::optional<c10::ArrayRef> size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use optional<ArrayRef> in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

swolchok commented:
@pytorchbot rebase

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the c10::optional<c10::ArrayRef> size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use optional<ArrayRef> in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

ezyang commented:
The errors look specific to torchvision, so at a guess you may need to update torchvision's use of optional simultaneously when you do this.

swolchok commented:
> you may need to update torchvision's use of optional simultaneously

is torchvision compiled separately such that we aren't recompiling the .so when we build our tests?

swolchok commented:
> is torchvision compiled separately such that we aren't recompiling the .so when we build our tests? this change is ABI-breaking.

(CC @malfet who might know?)

ezyang commented:
For CI, we should be building it from source.

cyyever commented:
@swolchok May be you can keep something like using c10::optional =std::optional in the old header until all downstream libs have changed to std::optional

swolchok commented:
> @swolchok May be you can keep something like using c10::optional =std::optional in the old header until all downstream libs have changed to use std::optional

that is how this diff is implemented already.

malfet commented:
> > is torchvision compiled separately such that we aren't recompiling the .so when we build our tests? this change is ABI-breaking.

@swolchok as Ed have said in the previous comment, torchvision is compiled from source in CI. Also, it sort of a red-herring. `test_custom_compound_op_autograd` will crash in the same fashion:
```
$ lldb -- python test_cpp_extensions_jit.py -v -k test_custom_compound_op_autograd
(lldb) target create "python"
Current executable set to 'python' (x86_64).
(lldb) settings set -- target.run-args  "test_cpp_extensions_jit.py" "-v" "-k" "test_custom_compound_op_autograd"
(lldb) r
Process 3794 launched: '/opt/conda/envs/py_3.8/bin/python' (x86_64)
Process 3794 stopped and restarted: thread 1 received signal: SIGCHLD
No CUDA runtime is found, using CUDA_HOME='/usr/local/cuda'
Test results will be stored in test-reports/python-unittest/test_cpp_extensions_jit

Running tests...
----------------------------------------------------------------------
  test_custom_compound_op_autograd (__main__.TestCppExtensionJIT) ... warning: Using /var/lib/jenkins/.cache/torch_extensions/py38_cpu as PyTorch extensions root...
Creating extension directory /var/lib/jenkins/.cache/torch_extensions/py38_cpu/is_python_module...
Emitting ninja build file /var/lib/jenkins/.cache/torch_extensions/py38_cpu/is_python_module/build.ninja...
Building extension module is_python_module...
Using envvar MAX_JOBS (6) as the number of workers...
[1/2] c++ -MMD -MF main.o.d -DTORCH_EXTENSION_NAME=is_python_module -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_clang\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1002\" -isystem /opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/include -isystem /opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/include/torch/csrc/api/include -isystem /opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/include/TH -isystem /opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/include/THC -isystem /opt/conda/envs/py_3.8/include/python3.8 -D_GLIBCXX_USE_CXX11_ABI=1 -fPIC -std=c++17 -c /var/lib/jenkins/.cache/torch_extensions/py38_cpu/is_python_module/main.cpp -o main.o 
[2/2] c++ main.o -shared -L/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/lib -lc10 -ltorch_cpu -ltorch -ltorch_python -o is_python_module.so
Loading extension module is_python_module...
Process 3794 stopped
* thread #1, name = 'python', stop reason = signal SIGABRT
    frame #0: 0x00007ffff705ce87 libc.so.6`__GI_raise(sig=2) at raise.c:51
(lldb) bt
* thread #1, name = 'python', stop reason = signal SIGABRT
  * frame #0: 0x00007ffff705ce87 libc.so.6`__GI_raise(sig=2) at raise.c:51
    frame #1: 0x00007ffff705e7f1 libc.so.6`__GI_abort at abort.c:79
    frame #2: 0x00007fffdd22c35a libstdc++.so.6`__cxxabiv1::__terminate(void (*)()) + 10
    frame #3: 0x00007fffdd22c3c5 libstdc++.so.6`std::terminate() + 21
    frame #4: 0x00007fffdd22c658 libstdc++.so.6`__cxa_throw + 72
    frame #5: 0x00007fffdd610296 libc10.so`c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 246
    frame #6: 0x00007fffdd6104cb libc10.so`c10::detail::torchInternalAssertFail(char const*, char const*, unsigned int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 59
    frame #7: 0x00007fffdead027a libtorch_cpu.so`torch::Library::Library(torch::Library::Kind, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::optional<c10::DispatchKey>, char const*, unsigned int) + 1018
    frame #8: 0x00007fffcd4bfecf is_python_module.so`torch::detail::TorchLibraryInit::TorchLibraryInit(torch::Library::Kind, void (*)(torch::Library&), char const*, std::optional<c10::DispatchKey>, char const*, unsigned int) + 137
    frame #9: 0x00007fffcd4bc6c7 is_python_module.so`__static_initialization_and_destruction_0(int, int) + 166
    frame #10: 0x00007fffcd4bc716 is_python_module.so`_GLOBAL__sub_I_main.cpp + 19
...
```

malfet commented on aten/src/ATen/core/library.cpp:
```diff
@@ -54,7 +54,7 @@ CppFunction::~CppFunction() = default;
   : kind_(kind)
   , ns_(ns == "_" ? c10::nullopt : c10::make_optional(std::move(ns)))
-  , dispatch_key_((!k.has_value() || *k == c10::DispatchKey::CatchAll) ? c10::nullopt : k)
+  , dispatch_key_((!k.has_value() || *k == c10::DispatchKey::CatchAll) ? c10::optional<c10::DispatchKey>() : k)
```
Considering where the crash happens, this change looks suspicious, though, to be frank,  I'm not sure what's the difference between `std::optional<>()` and `std::optional<>(std::nullopt)` constructors.

cyyever replied:
Why not use redefining c10::nullopt to std::nullopt?

malfet replied:
Anyway, ignore that, the crash is due to some ABI issues between clang and gcc

swolchok replied:
> Why not redefine c10::nullopt to std::nullopt?
already done

ezyang commented:
@malfet What do you suggest we do? If std::optional is implemented differently on different compilers, we don't really have any choice except to force people to use exactly the same compiler.

malfet commented:
@ezyang implementation is the same (i.e. gcc and clang use the same headers when compiling the code), it's just compiler flags are different. My hunch is `-std=gnu++17` (used to compile torch) vs `-std=c++17` (used to compile extensions) or `-fabi-version=`.

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

malfet commented:
If one rebases (which Scott already did) past https://github.com/pytorch/pytorch/pull/103350 it should work, as in CI we would no longer compile PyTorch with clang, but tests/domains with gcc

malfet commented:
@pytorchbot rebease

malfet commented:
@pytorchbot rebase

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

malfet committed:
Update test.sh

Testing whether this will fix things

malfet commented:
@pytorchbot rebase

swolchok committed:
rebase on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

swolchok commented:
@pytorchbot rebase

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

swolchok commented:
I thought the optTypeMetaToScalarType issue was on trunk, but it's ADL biting us again. Will fix.

swolchok committed:
fix lints and fix optTypeMetaToScalarType ADL on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan EikanWang jgong5

[ghstack-poisoned]

cyyever commented:
Still have some optTypeMetaToScalarType issues

swolchok committed:
fix another optTypeMetaToScalarType ADL on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

cyyever commented:
bugprone-unchecked-optional-access becomes annoying, I suggest disable it in clang-tidy config

swolchok committed:
suppress incorrect clang-tidy warnings on TorchDispatchModeTLS.cpp on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

swolchok commented:
@pytorchbot rebase

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

swolchok commented:
no idea why XLA is failing with "ModuleNotFoundError: No module named 'torch.version'"

cyyever commented:
@pytorchbot rebase

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

malfet commented:
@pytorchbot rebase

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

swolchok committed:
tons of lint suppressions on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

swolchok commented:
looks like we're just waiting for the next executorch pinned commit bump

cyyever commented:
The test failure may be unrelated and can be ignored.

swolchok commented:
> The test failure may be unrelated and can be ignored.

it's definitely related. https://github.com/pytorch/executorch/commit/61aad492a86aa8dcd6ce4764b3fac1a2ebb08520 used to be part of this PR but I had to split it out, and we need it pulled in so that this PR won't break the build.

#114717 will update us.

malfet commented:
@pytorchbot rebase -b main

malfet commented:
Considering that all other tests were green, we just need to wait for executorch tests to pass to merge this

swolchok committed:
Update on "[PyTorch] Redirect c10::optional to std::optional"


We have C++17 now!

I am intentionally dropping the `c10::optional<c10::ArrayRef>` size optimization. It was intended to improve dispatch, but thanks to D34602980 / #70864 we don't use `optional<ArrayRef>` in function arguments anymore anyway.

Differential Revision: [D46079028](https://our.internmc.facebook.com/intern/diff/D46079028/)

cc ezyang gchanan jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

malfet commented:
@pytorchbot merge

malfet commented:
@pytorchbot merge

malfet commented:
@pytorchbot merge

malfet commented:
@pytorchbot merge -f "Before rebase everything was green but executorch, see https://hud.pytorch.org/pytorch/pytorch/pull/101995?sha=18faa09012ed31baa8c37cce184dcbe36202d9ba "
